src/librustc_mir/transform/nll/infer.rs

```diff
@@ -0,0 +1,187 @@
+    where
+        'gcx: 'tcx + 'a,
+        'tcx: 'a,
+    {
+        let mut changed = true;
+        let mut dfs = Dfs::new(infcx);
+        while changed {
+            changed = false;
+            for constraint in &self.constraints {
+                let sub = &self.definitions[constraint.sub].value.clone();
+                let sup_def = &self.definitions[constraint.sup];
+                debug!("constraint: {:?}", constraint);
+                debug!("    sub (before): {:?}", sub);
+                debug!("    sup (before): {:?}", sup_def.value);
+
+                if dfs.copy(sub, &mut sup_def.value, constraint.point) {
+                    changed = true;
+                    if sup_def.capped {
+                        // This is kind of a hack, but when we add a
+                        // constraint, the "point" is always the point
+                        // AFTER the action that induced the
+                        // constraint. So report the error on the
+                        // action BEFORE that.
+                        assert!(constraint.point.statement_index > 0);
+                        let p = Location {
+                            block: constraint.point.block,
+                            statement_index: constraint.point.statement_index - 1,
+                        };
+
+                        self.errors.push(InferenceError {
+                            constraint_point: p,
+                            name: sup_def.name,
+                        });
+                    }
+                }
+
+                debug!("    sup (after) : {:?}", sup_def.value);
+                debug!("    changed     : {:?}", changed);
+            }
+            debug!("\n");
+        }
+
+        // TODO(nashenas88) this was in the prototype, but doesn't this clear
+        // out all of the errors???
+        mem::replace(&mut self.errors, IndexVec::new());
+    }
+}
+
+struct Dfs<'a, 'gcx: 'tcx + 'a, 'tcx: 'a> {
+    infcx: InferCtxt<'a, 'gcx, 'tcx>,
+}
+
+impl<'a, 'gcx: 'tcx + 'a, 'tcx: 'a> Dfs<'a, 'gcx, 'tcx> {
+    fn new(infcx: InferCtxt<'a, 'gcx, 'tcx>) -> Self {
+        Self { infcx }
+    }
+
+    fn copy(
+        &mut self,
+        from_region: &Region,
+        to_region: &mut Region,
+        start_point: Location,
+    ) -> bool {
+        let mut changed = false;
+
+        let mut stack = vec![];
+        let mut visited = FxHashSet();
+
+        stack.push(start_point);
+        while let Some(p) = stack.pop() {
+            debug!("        dfs: p={:?}", p);
+
+            if !from_region.may_contain(p) {
+                debug!("            not in from-region");
+                continue;
+            }
+
+            if !visited.insert(p) {
+                debug!("            already visited");
+                continue;
+            }
+
+            changed |= to_region.add_point(p);
+
+            let successor_points = self.infcx.tcx.successor_points(p);
```

Comment:
to obtain the successor points, we will need to look at the `Location`. We can load the basic block information [from the MIR via `basic_blocks()[location.block]`](https://github.com/rust-lang/rust/blob/72d65019c789138f555c7cf7139508d2f9f0dffe/src/librustc/mir/mod.rs#L169). Each block has a [`BasicBlockData`](https://github.com/rust-lang/rust/blob/72d65019c789138f555c7cf7139508d2f9f0dffe/src/librustc/mir/mod.rs#L560).

If we compare the `location.statement_index` and see that it less than `block_data.statements.len()`, then there is just one successor: `Location { statement_index: location.statement_index + 1, ..location }`.

Otherwise, `location.statement_index` should be equal to `block_data.statements.len()`, which indicates that this location is the "terminator" -- that is, the final point in the block. In that case, the successors can be found by invoking `block_data.terminator().successors()` to get a list of successor blocks. Those can be converted into `Location` by pairing the `BasicBlock` with a statement index of 0 (first statement in the block.